### PR TITLE
Make description toggleable

### LINKED ISSUE #20: Make description exapand/collapsible
Forecast descriptions have a lot of text! It'd be great if `short_description` could be displayed, and turn into `extended_description` on a click.

### PR DESCRIPTION
**Pull request by k-obrien**
Allow the user to toggle between the short and long descriptions when both are available.

Resolves #20

### CODE REVIEW

File: app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt [hunk 1]
```diff
@@ -196,6 +203,47 @@ private fun SectionEntry(@DrawableRes iconRes: Int, iconDesc: String, content: S
...
+@Composable
+private fun Description(
+  description: Description,
+  onToggle: (Boolean) -> Unit
+) {
+  val (text, expandable, expanded) = description
+
+  Column(
+    modifier = Modifier
+      .padding(start = 16.dp, end = 16.dp, bottom = 16.dp)
```
Comment (chris-horner) [inline app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt]: I think this might feel a bit nicer if we shift the padding declaration below the `clickable`, letting the touch ripple expand to the screen edges

File: app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt [hunk 2]
```diff
@@ -196,6 +203,47 @@ private fun SectionEntry(@DrawableRes iconRes: Int, iconDesc: String, content: S
...
+    Text(
+      text = text,
+      style = MaterialTheme.typography.body1,
+      modifier = Modifier.animateContentSize()
+    )
+
+    if (expandable) {
+      val rotation: Float by animateFloatAsState(if (expanded) 0F else -360F)
+      Icon(
+        imageVector = if (expanded) Icons.Rounded.ExpandLess else Icons.Rounded.ExpandMore,
```
Comment (chris-horner) [inline app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt]: We can simplify this by not swapping out the icon.

```suggestion
      val rotation: Float by animateFloatAsState(if (expanded) 180f else 0f)
      Icon(
        imageVector = Icons.Rounded.ExpandMore,
```

File: app/src/main/kotlin/codes/chrishorner/socketweather/home/HomeViewModel.kt [hunk 3]
```diff
@@ -42,20 +46,28 @@ class HomeViewModel(
   private val timeFormatter = DateTimeFormatter.ofPattern("h a")
   private val scope = overrideScope ?: viewModelScope
 
-  val states: StateFlow<HomeState> = tickerFlow(10_000, emitImmediately = true)
-    .flatMapLatest { forecaster.states }
-    .combine(locationStore.savedSelections) { forecasterState, savedSelections ->
-      forecasterState.toHomeState(savedSelections)
-    }
+  private val showExtendedDescription: MutableStateFlow<Boolean> = MutableStateFlow(false)
```
Comment (chris-horner) [inline app/src/main/kotlin/codes/chrishorner/socketweather/home/HomeViewModel.kt]: I'm looking to overhaul how viewmodels work in a bit. For now, I reckon for now this would be simpler if the expanded state was kept in the composable with a
```kotlin
var expanded by remember { mutableStateOf(false) }
```

That way we don't have to futz around with the viewmodel. It can still determine whether or not is `hasExtended`, but whether or not it's currently extended can be left as "view state" for now. (Though will likely change with some ideas I have in the future)

File: app/src/main/kotlin/codes/chrishorner/socketweather/home/HomeViewModel.kt [hunk 4]
```diff
@@ -65,13 +77,22 @@ class HomeViewModel(
...
+  private fun LoadingState.toHomeState(
+    savedSelections: Set<LocationSelection>,
+    showExtendedDescription: Boolean
+  ): HomeState {
     val toolbarTitle = when (this) {
-      Forecaster.LoadingState.Idle -> strings[R.string.home_loading]
-      is Forecaster.LoadingState.FindingLocation -> strings[R.string.home_findingLocation]
-      is Forecaster.LoadingState.Refreshing -> this.previousForecast.location.name
-      is Forecaster.LoadingState.Loaded -> this.forecast.location.name
-      is Forecaster.LoadingState.Error -> when (val selection = this.selection) {
```
Comment (chris-horner) [inline app/src/main/kotlin/codes/chrishorner/socketweather/home/HomeViewModel.kt]: Bunch of unrelated changes here that would be nice to avoid merging.

**Review by chris-horner — CHANGES_REQUESTED** [review]
Nice one! Just a couple of tweaks from me

File: app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt (hunk 1 above)
Comment (k-obrien) [reply app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt]: Good call. I've also updated the surrounding padding so that the ripple extends evenly below and above the description. There's no visible change otherwise regardless of the existence of a description.

_Description visible_
 

_Description hidden_
 

I'd have enabled view bounds for a better comparison but they don't seem to work for Compose. 🤔

File: app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt (hunk 2 above)
Comment (k-obrien) [reply app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt]: Nice one!

File: app/src/main/kotlin/codes/chrishorner/socketweather/home/HomeViewModel.kt (hunk 3 above)
Comment (k-obrien) [reply app/src/main/kotlin/codes/chrishorner/socketweather/home/HomeViewModel.kt]: That's actually my preference and I did it this way originally but using `rememberSaveable` because otherwise it forgets the expanded state on rotation; however, `rememberSaveable` doesn't appear to work (I couldn't figure out if I was doing something wrong but I did notice that the visible state of the location chooser uses it too and also doesn't survive rotation) so I opted for the VM route. Should I just use `remember` for now?

Comment (chris-horner) [reply app/src/main/kotlin/codes/chrishorner/socketweather/home/HomeViewModel.kt]: Yeah, I reckon we stick with remember for now 👍

File: app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt (hunk 1 above)
Comment (chris-horner) [reply app/src/main/kotlin/codes/chrishorner/socketweather/home/ForecastUi.kt]: Yeah, that's yet to be implemented, but I think it's coming!

**Review by chris-horner — APPROVED** [review]
Nice! Thanks for doing this

**Comment by k-obrien** [comment]
> Nice! Thanks for doing this

Happy to. It's giving me a place to learn Compose. 😅
